### PR TITLE
Calculate weights assigned to each grid point in DDEC6

### PR DESCRIPTION
**Pull request by mscho527**
_Related Issue: #885
Depends upon: #919_

This PR implements parts 1-4 of the fourth through seventh steps of DDEC6 algorithm. It uses the conditioned proatom densities and weighed average densities to perform stockholder type partition. Later PR will involve calculating quantities that update the densities based on this stockholder partition.

### CODE REVIEW

File: cclib/method/ddec.py [hunk 1]
```diff
@@ -590,9 +626,73 @@ def condition_densities(self):
...
-    def _phiai(self, ya, bigphiAI, atomi):
-        """ Evaluate phi_A^I based on equation S100
+    def calculate_w_and_u(self):
+        """ Calculate weights placed on each integration grid point
+            [STEP 4-7]
         """
+        # From equation 67, w_A(r_A) = self._cond_density
+        # From equation 8, W(r) = self.rho_cond
+        xshape, yshape, zshape = self.chgdensity.data.shape
+        atoms = len(self.data.atomnos)
```
Comment (berquist) [inline cclib/method/ddec.py:636]: `atoms` doesn't indicate that this is a count, prefer
```suggestion
        natoms = len(self.data.atomnos)
```

File: cclib/method/ddec.py [hunk 2]
```diff
@@ -590,9 +626,73 @@ def condition_densities(self):
             # Make tau monotonic decreasing
             self.tau[atomi] = numpy.maximum.accumulate(self.tau[atomi][::-1])[::-1]
 
-    def _phiai(self, ya, bigphiAI, atomi):
-        """ Evaluate phi_A^I based on equation S100
+    def calculate_w_and_u(self):
```
Comment (berquist) [inline cclib/method/ddec.py:629]: I think this method should be private.

File: cclib/method/ddec.py [hunk 3]
```diff
@@ -274,6 +274,38 @@ def calculate(self, indices=None, fupdate=0.05):
...
+        # Optimize weights on each grid point
+        kappa = [0.0 for x in self.data.atomnos]
+        update_kappa = False
+        self.N_A = []
+        steps = 4
+        while steps < 5:  # change this to 8 after implementing kappa routine
+            self.logger.info("Optimizing charge weights (Step {}).".format(steps))
+            self.N_A.append(self.calculate_w_and_u())
+
+            # Update kappa as described in (4) of doi: 10.1039/c6ra04656h
```
Comment (berquist) [inline cclib/method/ddec.py:287]: (4) doesn't look like the right reference.

File: cclib/method/ddec.py (hunk 1 above)
Comment (mscho527) [reply cclib/method/ddec.py]: Yes. I've made the necessary changes. Thanks!

File: cclib/method/ddec.py (hunk 2 above)
Comment (mscho527) [reply cclib/method/ddec.py]: I've made this private. Thanks!

File: cclib/method/ddec.py (hunk 3 above)
Comment (mscho527) [reply cclib/method/ddec.py]: I've updated this to a correct reference. Thanks!

File: cclib/method/ddec.py [hunk 4]
```diff
@@ -274,24 +274,56 @@ def calculate(self, indices=None, fupdate=0.05):
...
+            if (
+                steps > 5
+                and update_kappa
+                and numpy.diff(self.N_A)[-1] < 1e-5
+                and numpy.diff(self.N_A)[-2] < 1e-5
+            ):
+                update_kappa = False
+                kappa = [0.0 for x in self.data.atomnos]
+
+            # TODO: EXIT LOOP HERE for steps = 7
```
Comment (shivupa) [inline cclib/method/ddec.py:299]: if you change the if statement below to `if update_kappa and steps < 7:` then it will just exit the loop if steps == 7

Comment (mscho527) [reply cclib/method/ddec.py:299]: I was trying to wait until the full implementation of steps 4-7 (with PR921) before I add the exit conditions here. The exit condition is present in PR921 now. Thanks!

**Pull request merged**